File: algorithms/2_medium/C/01267_count_servers_that_communicate/solution.c

__attribute__((no_sanitize("all")))
int countServers(int** grid, const int rows, const int* cols_data)
{
    assert(rows >= 1 && rows <= 250 && cols_data != NULL);
    const int cols = cols_data[0];
    assert(cols >= 1 && cols <= 250);

    int* buff = calloc((size_t)(rows + cols), sizeof(*buff));
    assert(buff != NULL);
    int* sum_rows = &buff[0];
    int* sum_cols = &buff[rows];
    int total = 0;

    for (int i = 0; i < rows; i += 1) {
        for (int j = 0; j < cols; j += 1) {
            if (grid[i][j] == 1) {
                total += 1;
                sum_rows[i] += 1;
                sum_cols[j] += 1;
            }
        }
    }

    for (int i = 0; i < rows; i += 1) {
        for (int j = 0; j < cols; j += 1) {
            if (grid[i][j] == 1 && sum_rows[i] == 1 && sum_cols[j] == 1) {
                total -= 1;
            }
        }
    }

/* cleanup: */
    free(buff);
    return total;
}

File: algorithms/2_medium/C/01267_count_servers_that_communicate/solution.rs
impl Solution {
    pub fn count_servers(grid: Vec<Vec<i32>>) -> i32 {
        let rows = grid.len();
        assert!(rows >= 1 && rows <= 250);
        let cols = grid[0].len();
        assert!(cols >= 1 && cols <= 250);

        let mut buff = vec![0_i32; rows + cols];
        let (mut sum_rows, mut sum_cols) = buff.split_at_mut(rows);
        let mut total = 0_i32;

        for i in 0..rows {
            for j in 0..cols {
                if grid[i][j] == 1 {
                    total += 1;
                    sum_rows[i] += 1;
                    sum_cols[j] += 1;
                }
            }
        }

        for i in 0..rows {
            for j in 0..cols {
                if grid[i][j] == 1 && sum_rows[i] == 1 && sum_cols[j] == 1 {
                    total -= 1;
                }
            }
        }

        total
    }
}
